### PR TITLE
Prometheus exporter

### PR DESCRIPTION
**Pull request by RonnyAnc**
**What this PR does / why we need it**:

This is an extension method over IApplicationBuilder to simply add an endpoint on your application which could easily work as a prometheus exporter instead of using Prometheus Gateway. It could work for teams that work with exporters rather than Prometheus Gateway.

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #_[issue number]_

No issue, I didn't know if I had to create an issue first or not.

**Special notes for your reviewer**: 

I didn't have clear if this feature could be interesting for these repository so I talked with @CarlosLanderas via Twitter about this.

Maybe the project `HealthChecks.Publisher.Prometheus` is ot the place fo this feature because of this is publishing nothing.

I added a section in the documentation but maybe it is not the right place, you will know better.

I simply added a code example in the documentation, I do not know if that is enough for this.

**Does this PR introduce a user-facing change?**:

Yes, as it provides a new IApplicationBuilder extension method to add a new endpoint in their aspnet core apps.

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [X] Code compiles correctly
- [X] Created/updated tests
- [X] Unit tests passing
- [X] End-to-end tests passing
- [X] Extended the documentation
- [X] Provided sample for the feature

### CODE REVIEW

**Comment by unaizorrilla** [comment]
Hi @RonnyAnc 

Thanks for contribute this, looks good and agree with change gateway by exporter!!

I try to review this tomorrow!!!

**Comment by CarlosLanderas** [comment]
@RonnyAnc, thanks for your contribution.!

As you say, the Publisher project is not the best place to add this code as all Publishers are internally called from Microsoft HealthChecks package and implement an interface (IHealthCheckPublisher). All our publisher packages are following a push model.

As this is a new response writer that exposes a prometheus endpoint to scrape metrics, maybe I would create a new HealthChecks.Prometheus.Middleware or HealthChecks.Prometheus.Endpoint.

I can help you with it if you add me as contributor in your branch :)

https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push

**Comment by CarlosLanderas** [comment]
#271

**Comment by RonnyAnc** [comment]
Thank you for the feedback, I would try to change those things tomorrow 💪 Happy to hear you agree with the feature 😋

**Comment by RonnyAnc** [comment]
Heey!

I have just pushed the change for having two different projects, the new project is called HealthChecks.Prometheus.Metrics (I was thinking on HealthChecks.Prometheus.Exporter too). I did some changes that I would like you check too: 

- The PrometheusResponseWriter was duplicated in the new project because of in the future could not be the same implementation.
- The alwaysReturn200 was parameter for this new ResponseWriter because I thought it is not needed in the exporter.
- The tests for the old ResponseWriter were moved back as the were before and similar suite for the new one using the extension methods was added.
- I put the version of this package as 3.1.1-review but I really do not know which version should we manage.

Thank you so much!

**Comment by zh6335901** [comment]
I can't wait this feature!

File: src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Diagnostics.HealthChecks;
+
+namespace HealthChecks.Prometheus.Metrics.Extensions {
+    public static class PrometheusHealthCheckMiddleware {
```
Comment (unaizorrilla) [inline src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs:5]: Can you fix style of brackets?

File: src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+        public static IApplicationBuilder UseHealthChecksPrometheusExporter(this IApplicationBuilder applicationBuilder, string metricsEndpointName = "/healthmetrics")
```
Comment (unaizorrilla) [inline src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs:6]: - Probably endpoint parameter name is better than metricsEndpointName ( is not the name, is the endpoint path)
- PathString will be also good idea

File: src/HealthChecks.Publisher.Prometheus/PrometheusResponseWriter.cs [hunk 3]
```diff
@@ -28,5 +28,9 @@ public static async Task WritePrometheusResultText(HttpContext context, HealthRe
                 await context.Response.WriteAsync(content, Encoding.UTF8);
             }
         }
+
+        public static async Task WritePrometheusResultText(HttpContext context, HealthReport report) {
```
Comment (unaizorrilla) [inline src/HealthChecks.Publisher.Prometheus/PrometheusResponseWriter.cs:32]: fix brackets style

File: src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            instance.WriteMetricsFromHealthReport(report);
+
+            using (var resultStream = CollectionToStreamWriter(instance.Registry))
+            {
+                var content = await new StreamContent(resultStream)
+                   .ReadAsStringAsync();
+
+                context.Response.ContentType = ContentType;
+
+                await context.Response.WriteAsync(content, Encoding.UTF8);
```
Comment (unaizorrilla) [inline src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs:23]: Well, works, and probably we accept this, but from the performance point use the BodyWriter is better here!

File: test/FunctionalTests/HealthChecks.Publisher.Prometheus/PrometheusResponseWriterTests.cs [hunk 5]
```diff
@@ -35,10 +35,8 @@ public async Task be_healthy_when_health_checks_are()
                 })
                 .Configure(app =>
                 {
-                    app.UseHealthChecks("/health", new HealthCheckOptions
-                    {
-                        ResponseWriter = (context, report) =>
-                            PrometheusResponseWriter.WritePrometheusResultText(context, report)
+                    app.UseHealthChecks("/health", new HealthCheckOptions {
```
Comment (unaizorrilla) [inline test/FunctionalTests/HealthChecks.Publisher.Prometheus/PrometheusResponseWriterTests.cs:38]: Fix brackets style

**Comment by RonnyAnc** [comment]
Hi, 

Sorry people, these two last days were impossible for me. Thank you for the review @unaizorrilla, I will try to update the PR tomorrow :) Sorry for the brackets thing, normally I used another configuration and I missed them.

Thank you so much!

**Comment by unaizorrilla** [comment]
Hi @RonnyAnc 

No worries!!!

File: src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs (hunk 2 above)
Comment (RonnyAnc) [reply src/HealthChecks.Prometheus.Metrics/Extensions/ApplicationBuilderExtensions.cs:6]: I agree, endpoint sounds better :)

File: src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs (hunk 4 above)
Comment (RonnyAnc) [reply src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs:23]: I knew nothing about Body Writer, could you recommend me some reading?

**Comment by RonnyAnc** [comment]
I have just resolved the merge conflict and fix the comments you mentioned @unaizorrilla. Sorry for the delay, these days were a bit hard for me 🙏

Thank you!

**Comment by unaizorrilla** [comment]
No worries!!! I Will try to merge asap!

File: src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs (hunk 4 above)
Comment (unaizorrilla) [reply src/HealthChecks.Prometheus.Metrics/PrometheusResponseWriter.cs:23]: https://www.stevejgordon.co.uk/using-the-bodyreader-and-bodywriter-in-asp-net-core-3-0

**Comment by unaizorrilla** [comment]
Hi @RonnyAnc 

Merged, thanks for contribute this.

A release build is on appveyor, when finish a new package AspNetCore.HealthChecks.Prometheus.Metrics with version 3.1.1 will be available on NuGet.

**Pull request merged**